wp-includes/service-workers.php

```diff
@@ -0,0 +1,104 @@
+<?php
+/**
+ * Service Worker functions.
+ *
+ * @since ?
+ *
+ * @package PWA
+ */
+
+/**
+ * Initialize $wp_service_workers if it has not been set.
+ *
+ * @global WP_Service_Workers $wp_service_workers
+ * @return WP_Service_Workers WP_Service_Workers instance.
+ */
+function wp_service_workers() {
+	global $wp_service_workers;
+	if ( ! ( $wp_service_workers instanceof WP_Service_Workers ) ) {
+		$wp_service_workers = new WP_Service_Workers();
+	}
+	return $wp_service_workers;
+}
+
+/**
+ * Register a new service worker.
+ *
+ * @since ?
+ *
+ * @param string          $handle Name of the service worker. Should be unique.
+ * @param string|callable $src    Callback method or relative path of the service worker.
+ * @param array           $deps   Optional. An array of registered script handles this depends on. Default empty array.
+ * @param array           $scopes Optional Scopes of the service worker.
+ * @return bool Whether the script has been registered. True on success, false on failure.
+ */
+function wp_register_service_worker( $handle, $src, $deps = array(), $scopes = array() ) {
+	$service_workers = wp_service_workers();
+	$registered      = $service_workers->register( $handle, $src, $deps, $scopes );
+
+	return $registered;
+}
+
+/**
+ * Get service worker URL by scope.
+ *
+ * @param string $scope Scope, for example 'wp-admin'.
+ * @return string Service Worker URL.
+ */
+function wp_get_service_worker_url( $scope ) {
+
+	if ( ! $scope ) {
+		$scope = site_url( '/', 'relative' );
+	}
+
+	return add_query_arg( array(
+		'wp_service_worker' => $scope,
+	), site_url( '/', 'relative' ) );
+}
+
+/**
+ * Print service workers' scripts.
+ */
+function wp_print_service_workers() {
+
+	$scopes = wp_service_workers()->get_scopes();
+	if ( empty( $scopes ) ) {
+		return;
+	}
+	?>
+	<script>
+		<?php
+		foreach ( $scopes as $scope ) {
+			?>
+			if ( navigator.serviceWorker ) {
+				if ( ! window.$serviceWorkersRegister ) {
```

Comment:
This variable should perhaps get namespaces under the `wp` object. For example, `wp.serviceWorkerRegistrations`.

Reply:
Would this also mean that we should wrap registering the service worker with ` window.addEventListener('load', function() {` for `wp` to be available? Actually according to [this document](https://developers.google.com/web/fundamentals/primers/service-workers/registration) this is the best practice anyway.

Reply:
I was thinking you'd just define the object of it isn't available yet. I believe that is what all scripts that interact with the `wp` global do. In other words, `if ( ! window.wp ) { window.wp = {}; }`

Reply:
Hummm, if it is a best practice to wait until `load` then how would other scripts get access to the registration promises?

Reply:
Wouldn't waiting until `load` within other scripts as well work?

Reply:
If this script runs first then I suppose so, yes. It would calling the `load` event handlers in the order they were registered, which is the spec'ed behavior I believe.

The drawback here I see being that it would require every script that wants to access the `ServiceWorkerRegistration` object would have to add a `load` event handler. Maybe that is not a bad idea. I suppose we could require that now. If we decide otherwise, then an alternative could be to create our own `Promise` which then, after `load` is happens and we call `navigator.serviceWorker.register()` and it resolves, we can then resolve the original promise we created during page load with the registration object that get get later. But that is perhaps overkill here.

Reply:
👉  Also, I wonder if we actually need `wp.serviceWorkerRegistrations` or if it is redundant. I see actually that there is [`navigator.serviceWorker.getRegistration()`](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerContainer/getRegistration) which seems like a built-in way to get the same thing?

Reply:
Made the other changes and added waiting until `load` for now -- considering that we might change it if it becomes necessary in the future.

Reply:
>Also, I wonder if we actually need wp.serviceWorkerRegistrations or if it is redundant. I see actually that there is navigator.serviceWorker.getRegistration() which seems like a built-in way to get the same thing?

Good point, testing.

Reply:
Looks like you're right -- the promise can be accessed via `navigator.serviceWorker.getRegistration()` instead.